### PR TITLE
Test action interaction with platform

### PR DESCRIPTION
**Pull request by sluongng**
Add a test to show how no-remote-exec exec prop work w. test action

The most recent test added demonstrated that if `no-remote-exec` is
added to a platform and action were to run on that platform, then that
action will be executed locally instead of remotely. The intended use
case for this was to setup a local "fallback" exec platform that
actions could run on if none of the registered exec platform could be
used. However, this does not hold with test action.

Each test target in Bazel is composed of 2 types of actions:

a. One to build the test executable binary
b. One to run the executable binary to obtain test logs and result

Actions in group (a) are executed on the matched `exec` platform and
produce the artifact intended for the `target` platform. This artifact
is then used in (b), which means that actions in (b) must be executed
in `target` platform, defined by `--platforms` flag.

Provide a test to demonstrate that currently, there is no way to tell
Bazel to execute this test action (b) in the local fallback platform.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/analysis/test/TestActionBuilder.java [hunk 1]
```diff
@@ -366,8 +364,7 @@ private TestParams createTestAction()
     ImmutableList.Builder<Artifact> coverageArtifacts = ImmutableList.builder();
     ImmutableList.Builder<ActionInput> testOutputs = ImmutableList.builder();
 
-    ActionOwner actionOwner =
-        testConfiguration.useTargetPlatformForTests() ? getTestActionOwner() : getOwner();
+    ActionOwner actionOwner = getTestActionOwner(testConfiguration.useTargetPlatformForTests());
```
Comment (sluongng) [inline src/main/java/com/google/devtools/build/lib/analysis/test/TestActionBuilder.java:367]: This was an unrelated refactoring I did while tracing @AustinSchuh's past work.
It's a no-op change, but it helped me trace the current logic related to execution requirements a bit easier.

I can drop the hunks if folks don't like it.

Comment (katre) [reply src/main/java/com/google/devtools/build/lib/analysis/test/TestActionBuilder.java:367]: I like pushing the logic down into `getTestActionOwner`.

**Review by katre — CHANGES_REQUESTED** [review]
The change and the new test looks good. I'm a little unclear on the description, however: is this PR adding a new test to verify that the current behavior does not change in the future, or is it showing that the current behavior is wrong (and thus if we fixed test platforms, we'd need to update the test)?

If it's the latter, please add a comment to the test explaining what should actually be happening and what the expected outcome is.

**Comment by sluongng** [comment]
> If it's the latter, please add a comment to the test explaining what should actually be happening and what the expected outcome is.

It is the latter.
I think my desired behavior is a combination of things:

- Having the ability to change a test action's execution platform to be different than the default target platform. In a big repo setup, you might want an additional `test_exec_compatible_with` so that you can bind the test action under some constraints.

- If the test action's exec platform has `exec_properties` being `no-remote` or `no-remote-exec`, they should be copied to it's execution info accordingly and reflect in `TestTargetProperties.isRemotable()` result. Effectively providing a way that we could configure the test action to be executed on the local machine.

It's unclear to me how the exec's platform properties should work with the `test.` exec_group's properties. I guess that the exec_group properties should override the platform properties. So the effect of `use_target_platform_for_tests` should be adjusted accordingly.

**Comment by katre** [comment]
This all makes sense. One of my low-level tasks in Q3 is to think about tests and platforms and at least start writing up requirements, so this is a good start.

If you can add comments to the test explaining that this is the current state but there's more control desired, I'd appreciate that, and then this can be merged.

**Comment by sluongng** [comment]
Done.

**Review by katre — APPROVED** [review]
Thanks!

**Comment by sluongng** [comment]
also cc: @Silic0nS0ldier who opened https://github.com/bazelbuild/bazel/discussions/20497 and created https://github.com/Silic0nS0ldier/bazel-proposals/blob/4917efb804054fd1c9700e127cad94a4bc0e4d6b/designs/2023-06-04-exec-platform-scoped-spawn-strategies.md with similar effect.

**Pull request closed without merging**